api/src/main/java/org/pmiops/workbench/institution/InstitutionServiceImpl.java

```diff
@@ -0,0 +1,146 @@
+  private final Provider<InstitutionEmailDomainDao> institutionEmailDomainDaoProvider;
+  private final Provider<InstitutionEmailAddressDao> institutionEmailAddressDaoProvider;
+
+  @Autowired
+  InstitutionServiceImpl(
+      Provider<InstitutionDao> institutionDaoProvider,
+      Provider<InstitutionEmailDomainDao> institutionEmailDomainDaoProvider,
+      Provider<InstitutionEmailAddressDao> institutionEmailAddressDaoProvider) {
+    this.institutionDaoProvider = institutionDaoProvider;
+    this.institutionEmailDomainDaoProvider = institutionEmailDomainDaoProvider;
+    this.institutionEmailAddressDaoProvider = institutionEmailAddressDaoProvider;
+  }
+
+  @Override
+  public List<Institution> getInstitutions() {
+    return StreamSupport.stream(institutionDaoProvider.get().findAll().spliterator(), false)
+        .map(this::toModelClass)
+        .collect(Collectors.toList());
+  }
+
+  @Override
+  public Optional<Institution> getInstitution(final String id) {
+    return getDbInstitution(id).map(this::toModelClass);
+  }
+
+  @Override
+  public Institution createInstitution(final Institution institutionToCreate) {
+    return toModelClass(saveInstitution(institutionToCreate, new DbInstitution()));
+  }
+
+  @Override
+  public boolean deleteInstitution(final String id) {
+    return getDbInstitution(id)
+        .map(
+            dbInst -> {
+              institutionDaoProvider.get().delete(dbInst);
+              return true;
+            })
+        .orElse(false);
+  }
+
+  @Override
+  public Optional<Institution> updateInstitution(
+      final String id, final Institution institutionToUpdate) {
+    return getDbInstitution(id)
+        .map(dbInst -> toModelClass(saveInstitution(institutionToUpdate, dbInst)));
+  }
+
+  private Optional<DbInstitution> getDbInstitution(String id) {
+    return institutionDaoProvider.get().findOneByShortName(id);
+  }
+
+  private DbInstitution saveInstitution(final Institution modelClass, final DbInstitution dbClass) {
+    dbClass.setShortName(modelClass.getShortName());
+    dbClass.setDisplayName(modelClass.getDisplayName());
+    dbClass.setOrganizationTypeEnum(
+        DbStorageEnums.organizationTypeToStorage(modelClass.getOrganizationTypeEnum()));
+    dbClass.setOrganizationTypeOtherText(modelClass.getOrganizationTypeOtherText());
+
+    // save so the domain and address DAOs have something to reference
+    institutionDaoProvider.get().save(dbClass);
+
+    final InstitutionEmailDomainDao domainDao = institutionEmailDomainDaoProvider.get();
+    domainDao.deleteAllByInstitution(dbClass);
+
+    Optional.ofNullable(modelClass.getEmailDomains())
+        .ifPresent(
+            domains -> {
+              Set<DbInstitutionEmailDomain> dbDomains =
+                  domains.stream()
+                      .map(domain -> new DbInstitutionEmailDomain(dbClass, domain))
+                      .collect(Collectors.toSet());
+              domainDao.save(dbDomains);
```

Comment:
May want to use the saved Iterable here as well, before setting email domains below.